File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int jumpSolve(int i, vector<int>& dp, vector<int>& nums) {
	if (i >= nums.size() - 1)return 0;
	if (~dp[i])return dp[i];
	int ans = 1e6;
	for (int j = nums[i]; j >=1 ; j--) {
		ans = min(ans, 1 + jumpSolve(i + j, dp, nums));
	}
	return dp[i] = ans;
}

int jump(vector<int>& nums) {
	vector<int>dp(nums.size(), -1);
	return jumpSolve(0, dp, nums);
}
};

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums)  {
       int[] dp = new int[nums.length];
       Arrays.fill(dp, Integer.MAX_VALUE);
       dp[0] = 0;
       for(int i=0;i<nums.length;i++){
           for(int j = 0; j <= nums[i] && i + j < nums.length;j++){
               int next = i + j;
               dp[next]=Math.min(dp[next], dp[i] + 1);
           }
       }
       return dp[nums.length - 1];
    }
}